core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -839,15 +841,73 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
             .update(err, state)
             .flatMap(
               decision =>
-                if (decision.cont) clock.sleep(decision.delay) *> loop(decision.state)
-                else orElse(err, decision.finish()).map(Left(_))
+                decision.delay.run.flatMap { delay =>
+                  if (decision.cont) clock.sleep(delay) *> loop(decision.state)
+                  else orElse(err, decision.finish()).map(Left(_))
+                }
             ),
         succ => ZIO.succeedRight(succ)
       )
 
     policy.initial.flatMap(loop)
   }
 
+  /**
+   * Schedule this effect using the given schedule.
+   * Scheduled effects will run only when schedule signal it can run, keeping it dormant while it can't or until it fails.
+   */
+  final def scheduled[R1 <: R, B](schedule: ZSchedule[R1, Unit, B]): ZIO[R1 with Clock, E, B] =
```

Comment:
How do these functions differ from `repeat` etc.?

Reply:
Repeat and retry have a semantic of running effect first before evaluating schedule. Scheduled instead evaluates first before running effect, possibly blocking operation since the beginning. Does it make sense?

Reply:
Maybe we should change `repeat` semantic to be `schedule`?

Reply:
Good question :smile: 
I can see both being used

Reply:
Here's my view:

If we change `repeat` to have semantic of `schedule`, then if you want _old_ behavior, you can just do: `effect *> effect.repeat(schedule)`.

Right?

If this is correct, it's a very good argument that new `schedule` is more foundational than the old `repeat`.

Reply:
Makes a lot of sense!

I need to get used to see things as compositional, instead offering a hard method with a hard functionality

I'm really learning a lot with this PR! :smiley:

Reply:
Just stumbled in a wall and want some help :)
Scheduled signature is:
`final def scheduled[R1 <: R, B](schedule: ZSchedule[R1, Unit, B]): ZIO[R1 with Clock, E, B]`

It means it receives only schedules doesn't need any external input to calculate next state. It's a bit restrictive, right?

But without Unit, I can't see how to block since the beginning.

How can i solve this?